pkg/federate/kubefed/cluster.go

```diff
@@ -1,9 +1,314 @@
+	}
+
+	kubeConfig, err := f.buildFederatedClusterConfig(newCluster)
+	if err != nil {
+		klog.Errorf("Error building kube config for federated cluster %#v: %v", newCluster, err)
+		return
+	}
+
+	klog.Infof("KubeFedCluster \"%s\" has been updated", newCluster.Name)
+
+	f.Lock()
+	defer f.Unlock()
+
+	if f.clusterMap[newCluster.Name] == nil {
+		klog.Warningf("OnUpdate - no cached entry for KubeFedCluster \"%s\" exists", newCluster.Name)
+		return
+	}
+
+	f.clusterMap[newCluster.Name] = kubeConfig
+
+	for _, clusterWatcher := range f.clusterWatchers {
+		clusterWatcher.onUpdate(newCluster.Name, kubeConfig)
+	}
+}
+
+func (f *federator) buildFederatedClusterConfig(kubeFedCluster *fedv1.KubeFedCluster) (*rest.Config, error) {
+	apiEndpoint := kubeFedCluster.Spec.APIEndpoint
+	if apiEndpoint == "" {
+		return nil, fmt.Errorf("the API endpoint is empty")
+	}
+
+	secretName := kubeFedCluster.Spec.SecretRef.Name
+	if secretName == "" {
+		return nil, fmt.Errorf("the cluster does not have a secret name")
+	}
+
+	secret := &corev1.Secret{}
+	if err := f.kubeFedClient.Get(context.TODO(), client.ObjectKey{Namespace: kubeFedNamespace, Name: secretName}, secret); err != nil {
+		return nil, fmt.Errorf("error getting Secret \"%s\": %v", secretName, err)
+	}
+
+	token, tokenFound := secret.Data[corev1.ServiceAccountTokenKey]
+	if !tokenFound || len(token) == 0 {
+		return nil, fmt.Errorf("the secret %#v is missing a non-empty value for %q", secret, corev1.ServiceAccountTokenKey)
+	}
+
+	kubeConfig, err := clientcmd.BuildConfigFromFlags(apiEndpoint, "")
+	if err != nil {
+		return nil, err
+	}
+
+	kubeConfig.CAData = kubeFedCluster.Spec.CABundle
+	kubeConfig.BearerToken = string(token)
+
+	return kubeConfig, nil
+}
+
+func (f *federator) initKubeFedClusterInformer(listerWatcher cache.ListerWatcher) {
+	// Providing 0 duration to an informer indicates that resync should be delayed as long as possible
+	resyncPeriod := 0 * time.Second
+	_, informer := cache.NewInformer(listerWatcher, &fedv1.KubeFedCluster{}, resyncPeriod, f)
+	f.kubeFedClusterInformer = informer
+}
+
+func newKubeFedClusterListerWatcher(kubeFedConfig *rest.Config) (cache.ListerWatcher, error) {
+	var obj runtime.Object = &fedv1.KubeFedCluster{}
+
+	gvk, err := apiutil.GVKForObject(obj, scheme.Scheme)
+	if err != nil {
+		return nil, fmt.Errorf("error finding GroupVersionKind for %T: %v", obj, err)
+	}
+
+	klog.V(3).Infof("GroupVersionKind for %T: %#v", obj, gvk)
+
+	mapper, err := apiutil.NewDiscoveryRESTMapper(kubeFedConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error creating RESTMapper from config: %v", err)
+	}
+
+	mapping, err := mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+	if err != nil {
+		return nil, fmt.Errorf("error creating RESTMapping for %#v: %v", gvk, err)
+	}
+
+	client, err := apiutil.RESTClientForGVK(gvk, kubeFedConfig, scheme.Codecs)
+	if err != nil {
+		return nil, fmt.Errorf("error creating RESTClient for %#v: %v", gvk, err)
+	}
+
+	return &cache.ListWatch{
```

Comment:
I'm pretty sure I put a comment about this before and it disappeared after the push....

It is much easier to read and test if ListFunc and WatchFunc are not anonymous functions that are coded in-place.

Reply:
I had answered your previous comment. The functions are now only 2 -3 lines long - I think it's overkill to make them standalone and I don't see how making it inline affects readability. It's common in Go to inline small functions.  And as I mentioned before, it would not be useful or simple to try to UT `client.Get()`.